sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -5816,20 +5816,37 @@ pi_result piEnqueueEventsWaitWithBarrier(pi_queue Queue,
   std::scoped_lock<ur_shared_mutex> lock(Queue->Mutex);
 
   // Helper function for appending a barrier to a command list.
-  auto insertBarrierIntoCmdList =
-      [&Queue](pi_command_list_ptr_t CmdList,
-               const _pi_ze_event_list_t &EventWaitList, pi_event &Event,
-               bool IsInternal) {
-        if (auto Res = createEventAndAssociateQueue(
-                Queue, &Event, PI_COMMAND_TYPE_USER, CmdList, IsInternal))
-          return Res;
+  auto insertBarrierIntoCmdList = [&Queue](
+                                      pi_command_list_ptr_t CmdList,
+                                      const _pi_ze_event_list_t &EventWaitList,
+                                      pi_event &Event, bool IsInternal) {
+    if (auto Res = createEventAndAssociateQueue(
+            Queue, &Event, PI_COMMAND_TYPE_USER, CmdList, IsInternal))
+      return Res;
 
-        Event->WaitList = EventWaitList;
-        ZE_CALL(zeCommandListAppendBarrier,
-                (CmdList->first, Event->ZeEvent, EventWaitList.Length,
-                 EventWaitList.ZeEventList));
-        return PI_SUCCESS;
-      };
+    Event->WaitList = EventWaitList;
+
+    // For in-order queue we don't need a real barrier, just add a "barrier"
+    // event signal because it is already guaranteed that previous commands
+    // are completed when the signal is started.
+    if (Queue->isInOrderQueue()) {
```

Comment:
could we add an env var? something like UR_L0_IN_ORDER_BARRIER=0 (adds appendSignalEvent) =1 (adds appendBarrier). this could help us evaluate which one is better.

Reply:
I think it is too low-level detail to get users control of it. We will measure the impact by just seeing diff from this change,